frontends/concrete-python/concrete/fhe/mlir/processors/assign_bit_widths.py

```diff
@@ -88,13 +88,38 @@ def max_encrypted_bitwidth_node(node: Node):
         return normal_p + 1
 
     if name in MULTIPLY and all(value.is_encrypted for value in node.inputs):
-        return normal_p + 1
+        # For operations that use multiply, an additional bit
+        # needs to be added to the bitwidths of the inputs.
+        # For single precision circuits the max of the input / output
+        # precisions will be taken in required_encrypted_bitwidth. For
+        # multi-precision, the circuit partitions will handle the
+        # input and output precisions separately.
+        all_inp_bitwidths = []
+        # Need a loop here to allow typechecking and make mypy happy
+        for inp in node.inputs:
+            dtype_inp = inp.dtype
+            assert isinstance(dtype_inp, Integer)
+            all_inp_bitwidths.append(dtype_inp.bit_width)
+
+        normal_p = max(all_inp_bitwidths)
+
+        # FIXME: This probably does not work well with multi-precision!
+        return max(normal_p + 1, node.output.dtype.bit_width)
 
     return normal_p
 
 
 def required_encrypted_bitwidth(nodes: Iterable[Node]) -> int:
-    """Give the minimal precision to implement all the nodes."""
+    """Give the minimal precision to implement all the nodes.
+
+    This function is called for both single-precision (for the whole circuit)
+    and for multi-precision circuits (for circuit partitions).
+
+    Ops for which the compiler introduces TLUs need to be handled explicitly
```

Comment:
Not sure about this comment. max_encrypted_bitwidth_node has no multiprecision logic for now.